crates/ibc/src/msg.rs

```diff
@@ -91,103 +90,86 @@ impl BorshDeserialize for MsgNftTransfer {
-/// address
-#[derive(Debug, Clone)]
-pub struct MsgAcknowledgement {
-    /// IBC acknowledgement message
-    pub message: IbcMsgAcknowledgement,
-    /// Shieleded transfer for MASP transaction
-    pub transfer: Option<ShieldingTransfer>,
-}
-
-impl BorshSerialize for MsgAcknowledgement {
-    fn serialize<W: std::io::Write>(
-        &self,
-        writer: &mut W,
-    ) -> std::io::Result<()> {
-        let encoded_msg = self.message.clone().encode_vec();
-        let members = (encoded_msg, self.transfer.clone());
-        BorshSerialize::serialize(&members, writer)
+impl From<IbcShieldingData> for String {
+    fn from(data: IbcShieldingData) -> Self {
+        HEXUPPER.encode(&data.serialize_to_vec())
     }
 }
 
-impl BorshDeserialize for MsgAcknowledgement {
-    fn deserialize_reader<R: std::io::Read>(
-        reader: &mut R,
-    ) -> std::io::Result<Self> {
-        use std::io::{Error, ErrorKind};
-        let (msg, transfer): (Vec<u8>, Option<ShieldingTransfer>) =
-            BorshDeserialize::deserialize_reader(reader)?;
-        let message = IbcMsgAcknowledgement::decode_vec(&msg)
-            .map_err(|err| Error::new(ErrorKind::InvalidData, err))?;
-        Ok(Self { message, transfer })
+/// Extract MASP transaction from IBC envelope
+pub fn extract_masp_tx_from_envelope(
+    envelope: &MsgEnvelope,
+) -> Option<MaspTransaction> {
+    match envelope {
+        MsgEnvelope::Packet(packet_msg) => match packet_msg {
+            PacketMsg::Recv(msg) => {
+                extract_masp_tx_from_packet(&msg.packet, false)
+            }
+            PacketMsg::Ack(msg) => {
+                extract_masp_tx_from_packet(&msg.packet, true)
+            }
+            PacketMsg::Timeout(msg) => {
+                extract_masp_tx_from_packet(&msg.packet, true)
+            }
+            _ => None,
+        },
+        _ => None,
     }
 }
 
-/// IBC timeout packet message with `Transfer` for refunding to a shielded
-/// address
-#[derive(Debug, Clone)]
-pub struct MsgTimeout {
-    /// IBC timeout message
-    pub message: IbcMsgTimeout,
-    /// Shieleded transfer for MASP transaction
-    pub transfer: Option<ShieldingTransfer>,
-}
-
-impl BorshSerialize for MsgTimeout {
-    fn serialize<W: std::io::Write>(
-        &self,
-        writer: &mut W,
-    ) -> std::io::Result<()> {
-        let encoded_msg = self.message.clone().encode_vec();
-        let members = (encoded_msg, self.transfer.clone());
-        BorshSerialize::serialize(&members, writer)
+/// Decode IBC shielding data from the memo string
+pub fn decode_masp_tx_from_memo(
+    memo: impl AsRef<str>,
+) -> Option<IbcShieldingData> {
+    let bytes = HEXUPPER.decode(memo.as_ref().as_bytes()).ok()?;
+    IbcShieldingData::try_from_slice(&bytes).ok()
+}
+
+/// Extract MASP transaction from IBC packet memo
+pub fn extract_masp_tx_from_packet(
+    packet: &Packet,
+    is_sender: bool,
+) -> Option<MaspTransaction> {
+    let is_ft_packet = if is_sender {
```

Comment:
What is `is_ft_packet` - whether or not the packet is for a fungible token? Why do we even need to know this?

Reply:
It needs to decode the packet. Unfortunately, the packet data structure isn't the same as the NFT packet data.
https://github.com/cosmos/ibc-rs/blob/7ff41b3804cda649e0d7f7c21389b8d65f8d9b4b/ibc-apps/ics20-transfer/types/src/packet.rs#L28-L33
https://github.com/cosmos/ibc-rs/blob/7ff41b3804cda649e0d7f7c21389b8d65f8d9b4b/ibc-apps/ics721-nft-transfer/types/src/packet.rs#L29-L43